### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by trojkat**

### CODE REVIEW

**Comment by tomislater** [comment]
:+1:

File: AUTHORS.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Authors
+
+This file contains the list of people involved in the development 
+of `querystringsafe_base64` along its history.
+
+* Michał Pawłowski <https://github.com/pawlowskimichal>
```
Comment (ghost) [inline AUTHORS.md]: My only active account is now: https://github.com/not-raspberry (I've got 3 of them :P ).

File: README.md [hunk 2]
```diff
@@ -1 +1,46 @@
...
+    '%3D'
+
+2. Some libraries tolerate the `=` in query string values:
+
+    >>> from urlparse import urlsplit, parse_qs
+
+    >>> parse_qs(urlsplit('http://aaa.com/asa?q=AAAA=BBBB=CCCC').query)
+    {'q': ['AAAA=BBBB=CCCC']}
+
+but the RFC 3986 underspecifies the query string so we cannot rely on `=` chars being handled by all publishers as it is done by urlparse.
```
Comment (ghost) [inline README.md]: Replace `publishers` with something more general (`web applications`?).

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    'pytest-cov',
+    'pylama',
+]
+
+
+setup(
+    name='querystringsafe_base64',
+    version=package_version,
+    author='Clearcode - The A Room',
+    author_email='thearoom@clearcode.cc',
```
Comment (ghost) [inline setup.py]: I'd do something like `'@'.join(['thearoom', 'clearcode.cc'])` to defend against email harvesters. But maybe my practices are a bit dated.

File: src/querystringsafe_base64/__init__.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
```
Comment (ghost) [inline src/querystringsafe_base64/__init__.py]: I don't like actual code (functions and classes) in `__init__.py`. To the point when I'm planning on writing a custom plugin for pylint for that.

If it fits in a single file, putting the code into 'src/querystringsafe_base64.py' would be better (because less nested).

File: src/querystringsafe_base64/__init__.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    Decode the result of querystringsafe_base64_encode or a regular base64.
+
+    .. note ::
+        As a regular base64 string does not contain dots, replcing dots with
+        equal signs does basically noting to it. Also,
+        base64.urlsafe_b64decode allows to decode both safe and unsafe base64.
+        Therefore this function may also be used to decode the regular base64.
+
+    :param (str, unicode) encoded: dot64 string or unicode
```
Comment (ghost) [inline src/querystringsafe_base64/__init__.py]: I forgot to replace `dot64` with `urlsafe base64` during the internal CR a year ago.

Now we have to replace all occurrences of `dot64` with `query-string-safe base64` (I'd use hyphens in text), as this is the name we decide to stick with. There are 2.

File: tests/test_encode_decode.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python
```
Comment (ghost) [inline tests/test_encode_decode.py:1]: There were more tests in the repo we're extracting the code from. These 2 seem to be written from scratch.

See: `tests/unit/test_base64dotted.py`. It's got what I believe is a good corner case coverage.

File: README.md [hunk 7]
```diff
@@ -1 +1,46 @@
...
+
+    >>> parse_qs(urlsplit('http://aaa.com/asa?q=AAAA=BBBB=CCCC').query)
+    {'q': ['AAAA=BBBB=CCCC']}
+
+but the RFC 3986 underspecifies the query string so we cannot rely on `=` chars being handled by all web applications as it is done by urlparse.
+
+Therefore we consider chars: `['+', '/', '=']` unsafe and we replace them with `['-', '_', '.']`. Characters `+` and `/` are already handled by `urlsafe_*` functions from base64 so only `=` is left for us. The `.` character has been chosen because it often appears in real world query strings and it is not used
+by base64.
+
+## Solution
```
Comment (ghost) [inline README.md]: Side note:

I would call this paragraph `## The solution` because we called the previous one `## The problem`. Sounds kind of more 'solid' :)

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+language: python
+python:
```
Comment (ghost) [inline .travis.yml:2]: 2.6 could also work.

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    url='https://github.com/ClearcodeHQ/querystringsafe_base64',
+    include_package_data=True,
+    install_requires=[],
+    tests_require=test_requires,
+    zip_safe=False,
+
+    license="LGPL",
+    classifiers=[
+        'Intended Audience :: Developers',
```
Comment (ghost) [inline setup.py:71]: `Environment :: Web Environment`

File: tests/test_encode_decode.py (hunk 6 above)
Comment (ghost) [inline tests/test_encode_decode.py:1]: :+1:

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  - 3.5
+  - pypy
+  - pypy3
+install:
+- python setup.py develop
+- pip install querystringsafe_base64[tests]
+- pip install coveralls
+script:
+- py.test --verbose --capture=no --showlocals --cov src/querystringsafe_base64
+- pylama
```
Comment (ghost) [inline .travis.yml:16]: It's missing `pylama.ini`.

```
[pylama]
linters = pep8,pep257,pyflakes,mccabe

[pylama:pep8]
max_line_length = 120

[pylama:*/__init__.py]
# W0611 - Ignore warnings about missing module-level docstrings in __init__.py files.
# D100 - Ignore 'imported but not used' warnings in __init__.py files.
ignore = W0611,D100
```

You can also add pylint. `pep257` will require you to fix new lines in docstrings. Have some fame :D

File: src/querystringsafe_base64.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    Decode the result of querystringsafe_base64_encode or a regular base64.
+
+    .. note ::
+        As a regular base64 string does not contain dots, replcing dots with
+        equal signs does basically noting to it. Also,
+        base64.urlsafe_b64decode allows to decode both safe and unsafe base64.
+        Therefore this function may also be used to decode the regular base64.
+
+    :param (str, unicode) encoded: dot64 string or unicode
```
Comment (ghost) [inline src/querystringsafe_base64.py]: `dot64` left here

File: src/querystringsafe_base64.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    :rtype: str
+    :return: decoded string
+    """
+    # If unicode comes in, strange TypeError is thrown from unicode.translate
+    # because it differs from str.translate. Base64 and dot64 are
```
Comment (ghost) [inline src/querystringsafe_base64.py]: And here.

File: .bumpversion.cfg [hunk 13]
```diff
@@ -0,0 +1,14 @@
+[bumpversion]
+commit = True
+tag = True
+message = "Release {new_version}"
+current_version = 0.1.0
```
Comment (fizyk) [inline .bumpversion.cfg:5]: always start with 0.0.0 ;)

Comment (tomislater) [inline .bumpversion.cfg:5]: Lol, I have always started with `0.0.1` :dancer:

Comment (fizyk) [inline .bumpversion.cfg:5]: 0.0.0 allows you to make any release you want later, 0.0.1 suggests it's already ready, and if you'd have errors.... :dancers:
